source/sema.h

```diff
@@ -19,23 +19,147 @@
 #define CPP2_SEMA_H
 
 #include "reflect.h"
+#include "dll.h"
 
 
 namespace cpp2 {
 
+static reflect_libraries* active_instance = nullptr;
+
+class reflect_libraries
+{
+    std::vector<error_entry>& errors;
+    std::string               in_progress = {};
+    std::vector<dll>          loaded      = {};
+
+    struct function_entry
+    {
+        std::variant<
+            std::monostate,
+            meta::mf_sign_type_in*,
+            meta::mf_sign_type_inout*
+        > type = {};
+        std::variant<
+            std::monostate,
+            meta::mf_sign_func_in*,
+            meta::mf_sign_func_inout*
+        > func = {};
+    };
+
+    std::map<std::string, function_entry> functions = {};
+
+public:
+    reflect_libraries(std::vector<error_entry>& errors_)
+        : errors{ errors_ }
+    { }
+
+    void load(std::string_view lib) {
+        active_instance = this;
+        in_progress = lib;
+        finally _{ [&](){ active_instance = nullptr; in_progress = ""; } };
+        auto& dl = loaded.emplace_back(in_progress);
+        if(!dl.is_open()) {
+            errors.emplace_back(
+                source_position(-1, -1),
+                "unable to load library '" + in_progress + "': " + dll::get_last_error()
+            );
+        }
+    }
+
+    bool try_apply(
+        [[maybe_unused]] std::string_view caller,
+        std::string const& name,
+        auto& decl
+    ) const {
+        // TODO(DyXel): Do full/better name look-up.
+        auto search = functions.find(name);
+        if(search == functions.end()) return false;
+        auto& entry = search->second;
+        auto visitor = [&](auto&& v) -> bool {
+            if constexpr(requires { v(decl); }) {
+                v(decl);
+                return true;
+            }
+            return false;
+        };
+        if constexpr(std::is_same_v<meta::type_declaration, CPP2_TYPEOF(decl)>)
+            return std::visit(visitor, entry.type);
+        else
+            return std::visit(visitor, entry.func);
```

Comment:
This is pretty involved. Why can you not call `entry.type` directly.

Would it be better to have two functions `try_apply_type` and `try_apply_function`. The lookup could be refactored into its own method and then the correct method could directly be called.

Reply:
like I mentioned, this code is throwaway but I personally don't feel like its involved. This code is very tame compared to the lambda overload trick I had before 😛 (visit is just a mess to use, what can I say).

Reply:
Ok. ;)